text/001-wasm-udfs.md

```diff
@@ -0,0 +1,348 @@
+```
+
+Output will also be messagepack-encoded, to decode we can use:
+
+```javascript
+// usage: node decode_output.js output
+const fs = require("fs");
+const msgpack = require("@msgpack/msgpack");
+console.log(msgpack.decode(fs.readFileSync(process.argv[2])));
+```
+
+## Feature wishlist
+
+These changes are not strictly necessary for supporting complex datatypes in
+UDFs, but would go a long way for the overall UDF experience.
+
+- The ability to replace (`CREATE OR REPLACE FUNCTION ...`) and delete
+  (`DROP FUNCTION ...`) UDFs. Not only is it wasteful to store old functions'
+  WASM modules indefinitely, it also forces users to rewrite queries which
+  depend on these functions when the function is modified, since the new version
+  will also need to have a new name.
+- The ability to load UDF WASM Modules from an HTTP URL instead of providing it
+  b64-ed as part of the JSON UDF metadata.
+- A first class `CREATE FUNCTION` statement syntax which doesn't pass arguments
+  as JSON, eg:
+
+      CREATE [OR REPLACE] FUNCTION concat2
+      RETURNING TEXT
+      [LANGUAGE wasiMessagePack]
+      MODULE 'https://...'
+      ENTRY 'concat2'`
+
+  The currently required `input_type` field may be omitted if Seafowl no longer
+  attempts to verify UDF arguments prior to invocation.
+
+## Key open decisions
+
+- Are we happy with the approach recommended in this RFC of replacing primitive
+  on-stack arguments and return values with serialized input and output for
+  UDFs? If so, are we happy with MessagePack as a serialization format?
+- Who is responsible for verifying the UDF type signatures against the type of
+  the selected tuple and expected response? With the bytestream-passing
+  approach, there is no guarantee that the UDF will return what it promises to
+  return in the `CREATE FUNCTION` statement.
+- How many `language` types do we want to support simultaneously? Should we
+  eventually deprecate `Wasm` if we introduce `wasiMessagePack`?
+- Which WASM languages should receive first-class support? Rust is pretty
+  obvious, but eg: C++ programmers may prefer a different MessagePack library
+  than C devs.
+- Performance considerations: How expensive is it to use WASI `stdin` and
+  `stdout`? If we used raw memory or WASI env vars, would that be faster? What's
+  the right performance vs simple dev experience tradeoff here?
+- UDF lifecycle: should modules' `_initialize` function be called before the
+  exported UDF function is invoked during `SELECT`s? If so, when? My take is
+  that we shouldn't do this: UDFs should be considered stateless.
+- Using MessagePack for encoding input and output opens the door to user-defined
+  aggregate functions, table functions and scalar functions with variable
+  argument count. Are these things features we want to leave open or even
+  support in our calling convention or should be ignore these use cases entirely
+  for now and introduce a new `language` when users need these things.
```

Comment:
We want to be able to extend the current convention without requiring people to migrate to a new language. I alluded to it in a comment on the "_Data serialization format_" line. Being able to accept and return a tuple of scalars / a single scalar, as well as a pair of (scalar, scalar) (current output + accumulator state for UDAFs) should do it. Note that by "scalar" I mean "any Arrow datatype", which also includes structs and lists.